controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -275,53 +276,51 @@ private ApplicationController applications() {
     private List<Job> computeReadyJobs(ApplicationId id) {
         List<Job> jobs = new ArrayList<>();
         applications().get(id).ifPresent(application -> {
-            List<Step> productionSteps = application.deploymentSpec().steps().stream()
-                                                    .filter(step -> step.deploysTo(prod) || step.zones().isEmpty())
-                                                    .collect(toList());
-
             Change change = application.changeAt(clock.instant());
             Optional<Instant> completedAt = max(application.deploymentJobs().statusOf(systemTest)
                                                         .<Instant>flatMap(job -> job.lastSuccess().map(JobRun::at)),
                                                 application.deploymentJobs().statusOf(stagingTest)
                                                         .<Instant>flatMap(job -> job.lastSuccess().map(JobRun::at)));
             String reason = "New change available";
-            List<Job> testJobs = Collections.emptyList();
+            List<Job> testJobs = null; // null means "uninitialised", while empty means "don't run any jobs".
 
             if (change.isPresent())
-                for (Step step : productionSteps) {
+                for (Step step : productionStepsOf(application)) {
                     Set<JobType> stepJobs = step.zones().stream().map(order::toJob).collect(toSet());
-                    Map<Instant, List<JobType>> jobsByCompletion = stepJobs.stream().collect(groupingBy(job -> completedAt(change, application, job).orElse(Instant.EPOCH)));
-                    if (jobsByCompletion.containsKey(Instant.EPOCH)) { // Step incomplete because some jobs remain, trigger those if the previous step was done, or required test steps.
-                        for (JobType job : jobsByCompletion.get(Instant.EPOCH)) {
+                    Map<Optional<Instant>, List<JobType>> jobsByCompletion = stepJobs.stream().collect(groupingBy(job -> completedAt(change, application, job)));
+                    if (jobsByCompletion.containsKey(Optional.empty())) { // Step is incomplete; trigger remaining jobs if ready, or their test jobs if untested.
+                        for (JobType job : jobsByCompletion.get(Optional.empty())) {
                             Versions versions = versions(application, change, deploymentFor(application, job));
                             if (isTested(application, versions)) {
-                                if (completedAt.isPresent()
-                                    && jobStateContains(application, job, idle)
+                                if (   completedAt.isPresent()
+                                    && jobStateIsAmong(application, job, idle)
```

Comment:
Consider extracting a method for this: `boolean canTrigger(job, application)`.

Reply:
Sure.

Reply:
Ok, I tried it out. I find it hinders readability, because the logic is hidden. It's also used just this once, so there isn't any code reuse benefit to it. Also, the argument list becomes quite large.